### PR TITLE
Add GSPO

### PR DESCRIPTION
**Pull request by faresobeid**
Added GSPO implementation 

---

Linear: Resolves PRIMERL-27

### CODE REVIEW

File: src/prime_rl/trainer/config.py [hunk 1]
```diff
@@ -203,7 +214,7 @@ class TrainerConfig(BaseSettings):
         Field(
             description="Whether to recompute the logprobs. If True, will always recompute logprobs and overwrite those found in the training batch.",
         ),
-    ] = True
+    ] = False
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/config.py]: should not be part of this pr

File: src/prime_rl/trainer/config.py [hunk 2]
```diff
@@ -118,14 +118,26 @@ class RatioConfig(BaseGRPOVariantConfig):
...
+    epsilon_low: Annotated[float, Field(ge=0)] = 4e-4
+    epsilon_high: Annotated[float, Field(ge=0)] = 4e-4
+
+class GSPORatioConfig(BaseGRPOVariantConfig):
+    """Configures the GSPO ratio loss."""
+
+    type: Literal["gspo_ratio"] = "gspo_ratio"
+    clip_ratio: Annotated[float, Field(ge=0)] = 1.1
+
+GRPOVariantsConfig: TypeAlias = Union[ClippingConfig, RatioConfig, GSPOClipConfig, GSPORatioConfig]
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/config.py]: this will need a slight refactor because of #643

Comment (mikasenghaas) [reply src/prime_rl/trainer/config.py]: should be very straight-forward tho

File: src/prime_rl/trainer/loss.py [hunk 3]
```diff
@@ -42,6 +42,29 @@ def grpo_loss(
             clip_ratio=grpo_loss_config.clip_ratio,
             highest_entropy_percentage=grpo_loss_config.highest_entropy_ratio_loss,
         )
+    elif isinstance(grpo_loss_config, GSPOClipConfig):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: this is also affected by #643

File: src/prime_rl/trainer/loss.py [hunk 4]
```diff
@@ -111,19 +134,96 @@ def grpo_loss_ratio(
...
+    advantages: Float[Tensor, "batch seq"],
+    original_logprobs: Float[Tensor, "batch seq"],
+    loss_mask: Int[Tensor, "batch seq"],
+    temperature: float,
+    epsilon_low: float,
+    epsilon_high: float,
+    highest_entropy_percentage: float,
+) -> tuple[Tensor, Tensor, Tensor]:
+    """
+    DeepSeek Math Loss: https://arxiv.org/abs/2402.03300
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: copy paste?

File: src/prime_rl/trainer/loss.py [hunk 5]
```diff
@@ -111,19 +134,96 @@ def grpo_loss_ratio(
...
+
+    Args:
+        policy_logprobs: Log probabilities from the policy model
+        ref_logprobs: Log probabilities from the reference model
+        advantages: Advantages for each token
+        beta: KL penalty coefficient
+        epsilon: Clipping parameter for PPO
+        ignore_index: Specifies a target value that is ignored and does not contribute to the loss
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: doesnt match args

File: src/prime_rl/trainer/loss.py [hunk 6]
```diff
@@ -111,19 +134,96 @@ def grpo_loss_ratio(
...
+
+    raw_ratio = ((per_token_logps - original_logprobs) * loss_mask).sum(dim=-1) / loss_mask.sum(dim=-1)
+    raw_ratio = raw_ratio.exp()
 
     ratio = torch.clamp(raw_ratio, 0, clip_ratio)
-    loss = -ratio * advantages
+    loss = -ratio * advantages[:, 0]
 
     if highest_entropy_percentage < 1.0:
         loss_mask = highest_entropy_mask(shifted_logits, loss_mask, highest_entropy_percentage)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: we removed this right?

File: src/prime_rl/trainer/loss.py [hunk 7]
```diff
@@ -111,19 +134,96 @@ def grpo_loss_ratio(
...
 
-    loss = _masked_sum(loss, loss_mask)
-    ratio = _masked_sum(ratio, loss_mask)
+    loss = loss.sum()
+    clipped_count = (raw_ratio > clip_ratio).float().sum()
+    raw_ratio = raw_ratio.sum()
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: it's a bit confusing imo that we call this raw_ratio again, maybe we can find smth more descriptive

File: src/prime_rl/trainer/train.py [hunk 8]
```diff
@@ -222,9 +222,10 @@ def train(config: TrainerConfig):
         batch_size = micro_batch_size * num_micro_batches
 
         # Normalize by the number of unmasked tokens in the batch (per-batch length normalization)
-        loss_scale = sum(micro_batch["loss_mask"].sum() for micro_batch in micro_batches)
+        entropy_scale = sum(micro_batch["loss_mask"].sum() for micro_batch in micro_batches)
+        loss_scale = entropy_scale if (config.loss.variant.type == "gspo_clip" or config.loss.variant.type == "gspo_ratio") else batch_size
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/train.py]: this changes the `loss_scale` compared for non-gspo right?

File: src/prime_rl/trainer/config.py (hunk 1 above)
Comment (faresobeid) [reply src/prime_rl/trainer/config.py]: ya just had set for running experiments

File: src/prime_rl/trainer/loss.py (hunk 4 above)
Comment (faresobeid) [reply src/prime_rl/trainer/loss.py]: oops

File: src/prime_rl/trainer/loss.py (hunk 5 above)
Comment (faresobeid) [reply src/prime_rl/trainer/loss.py]: oops

File: src/prime_rl/trainer/loss.py (hunk 7 above)
Comment (faresobeid) [reply src/prime_rl/trainer/loss.py]: good point, this change is just to log the importance ratios before clipping to be easier to debug/tune the clipping ratio. Should prob also have a different pr to log the max/min ratios

File: src/prime_rl/trainer/train.py (hunk 8 above)
Comment (faresobeid) [reply src/prime_rl/trainer/train.py]: oh oops this is wrong

File: src/prime_rl/trainer/train.py [hunk 9]
```diff
@@ -259,10 +260,10 @@ def train(config: TrainerConfig):
...
 
             # Accumulate unnormalized local metrics
-            loss_metrics["loss/loss"] += loss.detach().float()
-            loss_metrics["loss/entropy"] += entropy.detach().float()
-            loss_metrics["loss/importance_ratio"] += importance_ratio.detach().float()
-            loss_metrics["loss/clipped_ratio"] += clipped_token_count.detach().float()
+            loss_metrics["loss/loss"] += loss.detach().float() / loss_scale
+            loss_metrics["loss/entropy"] += entropy.detach().float() / entropy_scale
+            loss_metrics["loss/importance_ratio"] += importance_ratio.detach().float() / loss_scale 
+            loss_metrics["loss/clipped_ratio"] += clipped_token_count.detach().float() / loss_scale
```
Comment (samsja) [inline src/prime_rl/trainer/train.py]: why the change here ?

Comment (faresobeid) [reply src/prime_rl/trainer/train.py]: To log them properly

File: src/prime_rl/trainer/config.py [hunk 10]
```diff
@@ -119,7 +119,21 @@ class RatioLossConfig(BaseLossConfig):
...
+    epsilon_high: Annotated[float, Field(ge=0)] = 4e-4
+    clip_ratio: Annotated[float, Field(ge=0)] = 1.1
+
+class GSPORatioLossConfig(BaseLossConfig):
+    """Configures the GSPO ratio loss."""
+
+    type: Literal["ratio_gspo"] = "ratio_gspo"
+    clip_ratio: Annotated[float, Field(ge=0)] = 1.1
+
+LossConfig: TypeAlias = ClippingLossConfig | RatioLossConfig | GSPORatioLossConfig | GSPOClipLossConfig
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/config.py]: Let's call `ClippingLossConfig` -> `GRPOClipLossConfig` and `RatioLossConfig` -> `GRPORatioLossConfig` for consistency

File: src/prime_rl/trainer/config.py [hunk 11]
```diff
@@ -119,7 +119,21 @@ class RatioLossConfig(BaseLossConfig):
     clip_ratio: Annotated[float, Field(ge=0)] = 8.0
 
 
-LossConfig: TypeAlias = ClippingLossConfig | RatioLossConfig
+class GSPOClipLossConfig(BaseLossConfig):
+    """Configures the GSPO clip loss."""
+
+    type: Literal["clip_gspo"] = "clip_gspo"
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/config.py]: let's use `-` here instead of `_` (same as e.g. for advantage types)

Comment (mikasenghaas) [reply src/prime_rl/trainer/config.py]: so then the general types would be `grpo-clip`, `grpo-ratio`, `gspo-clip` and `gspo-ratio`

File: src/prime_rl/trainer/loss.py [hunk 12]
```diff
@@ -89,16 +111,76 @@ def grpo_loss_ratio(
...
+    shifted_logits: Float[Tensor, "batch seq vocab"],
+    input_ids: Int[Tensor, "batch seq"],
+    advantages: Float[Tensor, "batch seq"],
+    original_logprobs: Float[Tensor, "batch seq"],
+    loss_mask: Int[Tensor, "batch seq"],
+    temperature: float,
+    epsilon_low: float,
+    epsilon_high: float,
+    clip_ratio: float,
+) -> tuple[Tensor, Tensor, Tensor]:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py:136]: can we add a docstring here

File: src/prime_rl/trainer/loss.py [hunk 13]
```diff
@@ -89,16 +111,76 @@ def grpo_loss_ratio(
...
+@jaxtyped(typechecker=typechecker)
+def gspo_loss_ratio(
+    shifted_logits: Float[Tensor, "batch seq vocab"],
+    input_ids: Int[Tensor, "batch seq"],
+    advantages: Float[Tensor, "batch seq"],
+    original_logprobs: Float[Tensor, "batch seq"],
+    loss_mask: Int[Tensor, "batch seq"],
+    temperature: float,
+    clip_ratio: float,
+) -> tuple[Tensor, Tensor, Tensor]:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py:169]: and here

File: src/prime_rl/trainer/loss.py [hunk 14]
```diff
@@ -89,16 +111,76 @@ def grpo_loss_ratio(
...
+    
+    shifted_logits = shifted_logits / temperature
+    per_token_logps = selective_log_softmax(shifted_logits, input_ids)
+
+    raw_ratio = ((per_token_logps - original_logprobs) * loss_mask).sum(dim=-1) / loss_mask.sum(dim=-1)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: this seems to be the critical line, can you add some comments explaining the tensors shapes and how it translates to the math in the paper

File: src/prime_rl/trainer/train.py [hunk 15]
```diff
@@ -281,7 +282,10 @@ def train(config: TrainerConfig):
 
         # Normalize all loss metrics globally before reporting
         for key, value in loss_metrics.items():
-            loss_metrics[key] = value / loss_scale
+            if key == "loss/entropy":
+                loss_metrics[key] = value / total_tokens
+            else:
+                loss_metrics[key] = value / loss_scale
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/train.py]: argh this looks like code smell. seems like an easy footgun in the future. can we not move the normalization back in the grad accumulation loop? then the normalization factor is defined for each value explicitly which seems less likely to fail silently

**Comment by mikasenghaas** [comment]
Duplicate of #801. Reopen if mistake
